plutus-core/plutus-core/src/PlutusCore/Evaluation/Machine/ExMemory.hs

```diff
@@ -94,13 +96,36 @@ type CostingInteger =
     SatInt
 #endif
 
-
 -- $(if finiteBitSize (0::SatInt) < 64 then [t|Integer|] else [t|SatInt|])
 
+{- | Divide one costing integer by another, "rounding upwards".  This is needed
+when we expose costs to the ledger, which will use different (smaller) units.
+Suppose one ledger unit is 1000 real cost units: then if a script costs 12345678
+real units we want that to be converted to 12346 ledger units, since 12345 (as
+given by `div`) would convert to 12345000 real units, which wouldn't be enough
+to run the script.  We want a <= (a `uppperIntegerQuotient` b) * b < a+b for all
+a and for all b>0, except that we may get '==' instead of '<' when we're using
+SatInt and a+b == MaxBound.  The behaviour when b <= 0 is unspecified.
+-}
+uppperIntegerQuotient :: CostingInteger -> CostingInteger -> CostingInteger
+a `uppperIntegerQuotient` b =
+    let (q,r) = a `divMod` b
+    in if r==0 then q else q+1
+
+{- | A class which allows us to scale budget-related quantities upwards and
+   downwards by a given factor. -}
+class Scalable a where
```

Comment:
The actual numbers are wrapped inside a number of constructors and this seemed like the easiest way to get to them.

Reply:
I won't suggest we use https://hackage.haskell.org/package/monoid-extras-0.4.2/docs/Data-Monoid-Action.html#t:Action :trollface:

Reply:
We definitely shouldn't use that because what we're doing doesn't satisfy its axioms!  We don't have `(k1 <> k2) * c = k1 * (k2 * c)` because of the "rounding" upwards: that's kind of the whole point of what we're doing.

Reply:
That's true! The rounding does indeed make it not a proper action.